File: 240717/알파벳과 사칙연산/calculations-with-alphabet.js
function permutation(n){
    let arr = [[1], [2], [3], [4]]

    for (let i = 1; i < n; i++){
        new_arr = []
        for (let j = 1; j <= 4; j++){
            arr.forEach(elem => {
                tmp = elem.slice()
                tmp.push(j)
                new_arr.push(tmp)
            })
        }
        arr = new_arr
    }
    return arr
}

function calc(expression){
    const tokens = expression.split('');

    let result = Number(tokens[0]);

    for (let i = 1; i < tokens.length; i += 2) {
        const operator = tokens[i];
        const nextNumber = Number(tokens[i + 1]);

        if (operator === '+') {
            result += nextNumber;
        } else if (operator === '-') {
            result -= nextNumber;
        } else if (operator === '*') {
            result *= nextNumber;
        } else if (operator === '/') {
            result /= nextNumber;
        }
    }

    return result;
}

const fs = require("fs")
const input = fs.readFileSync(0).toString().trim()

const alp = new Map()
let idx = 0

for (let elem of input){
    if (elem.charCodeAt(0) >= 'a'.charCodeAt(0) && elem.charCodeAt(0) <= 'f'.charCodeAt(0)){
        if (!alp.has(elem)){
            alp.set(elem, idx)
            idx++
        }
    }
}

const perms = permutation(alp.size)
let ans = Number.MIN_SAFE_INTEGER

for (let perm of perms){
    exp = ""
    for (let elem of input){
        if (alp.has(elem)){
            exp += String(perm[alp.get(elem)])
        } else {
            exp += elem
        }
    }
    ans = Math.max(ans, calc(exp))
}

console.log(ans)

File: 240718/알파벳과 사칙연산/calculations-with-alphabet.py
def permutation(n):
    if n == 1:
        return [[1], [2], [3], [4]]
    tmp = []
    for i in permutation(n - 1):
        for j in range(1, 5):
            tmp.append(i + [j])
    return tmp

def calc(exp):
    result = int(exp[0])

    for i in range(2, len(exp), 2):
        num = int(exp[i])
        if exp[i - 1] == '+':
            result += num
        elif exp[i - 1] == '-':
            result -= num
        else:
            result *= num
    return result

expression = input()
idx = 0
alps = {}

for i in expression:
    if (ord('a') <= ord(i) <= ord('f') and i not in alps):
        alps[i] = idx
        idx += 1

result = float('-inf')
for perm in permutation(len(alps)):
    new_exp = ""
    for s in expression:
        if s in alps:
            new_exp += str(perm[alps[s]])
        else:
            new_exp += s
    result = max(result, calc(new_exp))

print(result)